src/rustzcash.rs

```diff
@@ -1019,3 +1021,493 @@ pub extern "system" fn librustzcash_sprout_verify(
         _ => false,
     }
 }
+
+pub struct SaplingProvingContext {
+    bsk: Fs,
+    bvk: edwards::Point<Bls12, Unknown>,
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_output_proof(
+    ctx: *mut SaplingProvingContext,
+    esk: *const [c_uchar; 32],
+    diversifier: *const [c_uchar; 11],
+    pk_d: *const [c_uchar; 32],
+    rcm: *const [c_uchar; 32],
+    value: uint64_t,
+    cv: *mut [c_uchar; 32],
+    zkproof: *mut [c_uchar; GROTH_PROOF_SIZE],
+) -> bool {
+    // Grab `esk`, which the caller should have constructed for the DH key exchange.
+    let esk = match Fs::from_repr(read_fs(&(unsafe { &*esk })[..])) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // Grab the diversifier from the caller.
+    let diversifier = sapling_crypto::primitives::Diversifier(unsafe { *diversifier });
+
+    // Grab pk_d from the caller.
+    let pk_d = match edwards::Point::<Bls12, Unknown>::read(&(unsafe { &*pk_d })[..], &JUBJUB) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // pk_d should be prime order.
+    let pk_d = match pk_d.as_prime_order(&JUBJUB) {
+        Some(p) => p,
+        None => return false,
+    };
+
+    // Construct a payment address
+    let payment_address = sapling_crypto::primitives::PaymentAddress {
+        pk_d: pk_d,
+        diversifier: diversifier,
+    };
+
+    // Initialize secure RNG
+    let mut rng = OsRng::new().expect("should be able to construct RNG");
+
+    // The caller provides the commitment randomness for the output note
+    let rcm = match Fs::from_repr(read_fs(&(unsafe { &*rcm })[..])) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // We construct ephemeral randomness for the value commitment. This
+    // randomness is not given back to the caller, but the synthetic
+    // blinding factor `bsk` is accumulated in the context.
+    let rcv = Fs::rand(&mut rng);
+
+    // Accumulate the value commitment randomness in the context
+    {
+        let mut tmp = rcv.clone();
+        tmp.negate(); // Outputs subtract from the total.
```

Comment:
Is this fine? Seems a lot like what you had to fix with an extra assignment.

Reply:
Ahh seems fine cause in `fs.rs` you change `self`

Reply:
sidenote:I'd consider changing the *other* `negate` methods in `edwards.rs` and `montgomery.rs` to mutate `self` or change the method names to e.g. `get_inverse`.  As we saw, it's  misleading a method called negate doesn't mutate the object :)